zkml/src/layers/transformer/embeddings.rs

```diff
@@ -81,12 +151,273 @@ impl<N: Number> Evaluate<N> for Embeddings<N> {
+    PCS: PolynomialCommitmentScheme<E>,
+{
+    type Proof = EmbeddingsProof<E>;
+
+    fn verify<T: Transcript<E>>(
+        &self,
+        proof: &Self::Proof,
+        last_claims: &[&Claim<E>],
+        verifier: &mut Verifier<E, T, PCS>,
+        _shape_step: &ShapeStep,
+    ) -> anyhow::Result<Vec<Claim<E>>> {
+        ensure!(
+            last_claims.len() == 1,
+            "embeddings only support 1 last claim"
+        );
+        // we verify the matmul proof first
+        let mut claims = self
+            .mat_ctx
+            .verify_matmul(verifier, last_claims[0], &proof.mat_proof)?;
+        ensure!(claims.len() == 1, "embeddings matmul should have 1 claim");
+        // the first claim is the one hot encoding claim. To verify it we need to
+        // efficiently evaluate the one hot encoding on it - we do this "at the end" of the verification
+        // procedure to respect the framework's order of operations. The logic is in `verify_input_claim`.
+        Ok(vec![claims.remove(0)])
+    }
+
+    fn verify_input_claim(&self, inputs: &[Tensor<E>], claims: &[&Claim<E>]) -> anyhow::Result<()> {
+        // TODO verify efficiently the one hot encoding claim
+        ensure!(inputs.len() == 1, "embeddings only support 1 input tensor");
+        ensure!(claims.len() == 1, "embeddings only support 1 claim");
+        let input = &inputs[0];
+        let one_hot_claim = &claims[0];
+        let vocab_nv = self.vocab_size.next_power_of_two().ilog2();
+        let seq_len_nv = input.get_shape().dim(0).next_power_of_two().ilog2();
+        ensure!(
+            vocab_nv + seq_len_nv == one_hot_claim.point.len() as u32,
+            "vocab_nv: {vocab_nv}, seq_len_nv: {seq_len_nv}, one_hot_claim.point.len(): {}",
+            one_hot_claim.point.len()
+        );
+        let (r1, r2) = one_hot_claim.point.split_at(seq_len_nv as usize);
+        let b1 = compute_betas_eval(r1);
+        let b2 = compute_betas_eval(r2);
+        let mut sum = E::ZERO;
+        for (idx, token) in input.get_data().iter().enumerate() {
+            let token_value = token.to_canonical_u64_vec()[0] as usize;
+            let selector = b1[idx] * b2[token_value];
+            sum += selector;
+        }
+        ensure!(
+            sum == one_hot_claim.eval,
+            "one hot encoding claim is incorrect"
+        );
+        Ok(())
+    }
+}
+
+fn one_hot_encoding<E: ExtensionField>(indices: &[E], vb: usize, mode: PaddingMode) -> Tensor<E> {
+    let mut data = Vec::new();
+    let vocab_size = match mode {
+        PaddingMode::NoPadding => vb,
+        PaddingMode::Padding => vb.next_power_of_two(),
+    };
+    for idx in indices {
+        let mut one_hot = vec![E::ZERO; vb];
+        let idx: usize = idx.to_canonical_u64_vec()[0].try_into().unwrap();
+        one_hot[idx] = E::ONE;
+        data.extend_from_slice(&one_hot);
+    }
+    let data = match mode {
+        PaddingMode::NoPadding => data,
+        PaddingMode::Padding => {
```

Comment:
Not super clear the padding logic here: we assume `indices.len` is already a power of 2, so if this is the case, then the previous loop should already produce a properly padded tensor, no? Or the purpose was padding with zero rows rather than one-hot encoded rows?